lib/Service/LoginService.php

```diff
@@ -119,6 +119,14 @@ public function login($profile, $userSession, $request)
             $uid = md5($uid);
         }
 
+        // Check if user is in allowed groups
+        if ($allowedGroups = $this->config->getSystemValue('oidc_login_allowed_groups', null)) {
+            $groupNames = $this->getGroupNames($profile);
+            if (empty(array_intersect($allowedGroups, $groupNames))) {
+                throw new LoginException($this->l->t('Access to Nextcloud is not allowed for this user (group)'));
```

Comment:
Maybe the message could be clearer?

*You cannot access to nextcloud because you are member of the allowed groups. If you think this is an error please contact your administrator.*

And the administrator email (the client contact attribute) could be displayed so users know who to contact.

Reply:
Done. I've no idea how to get the admin email so gonna defer that.